java/test/jmri/managers/AbstractManagerTestBase.java

```diff
@@ -109,11 +112,92 @@ public void vetoableChange(PropertyChangeEvent evt) throws PropertyVetoException
         Assert.assertEquals(base, l.getVetoableChangeListeners().length);
 
     }
+
     @Test
     public void testMakeSystemName() {
         String s = l.makeSystemName("1");
         Assert.assertTrue(s != null);
         Assert.assertTrue(! s.isEmpty());
     }
 
+    private Field getField(Class c, String fieldName) {
+        try {
+            return c.getDeclaredField(fieldName);
+        } catch (NoSuchFieldException ex) {
+            if (c.getSuperclass() != null)
+                return getField(c.getSuperclass(), fieldName);
+        }
+        
+        // Field not found
+        return null;
+    }
+    
+    @Test
+    public void testRegisterDuplicateSystemName() throws PropertyVetoException, NoSuchFieldException, NoSuchFieldException, IllegalArgumentException, IllegalAccessException {
+        if (l instanceof ProvidingManager)  {
+            ProvidingManager<E> m = (ProvidingManager<E>) l;
+            String s1 = l.makeSystemName("1");
+            String s2 = l.makeSystemName("2");
+            Assert.assertTrue(s1 != null);
+            Assert.assertTrue(! s1.isEmpty());
+            Assert.assertTrue(s2 != null);
+            Assert.assertTrue(! s2.isEmpty());
+
+            E e1;
+            E e2;
+
+            try {
+                e1 = m.provide(s1);
+                e2 = m.provide(s2);
+            } catch (IllegalArgumentException | NullPointerException | ArrayIndexOutOfBoundsException ex) {
+                // jmri.jmrix.openlcb.OlcbLightManagerTest gives a NullPointerException here.
+                // jmri.jmrix.openlcb.OlcbSensorManagerTest gives a ArrayIndexOutOfBoundsException here.
+                // Some other tests give an IllegalArgumentException here.
+
+                // If the test is unable to provide a named bean, abort this test.
+                JUnitAppender.clearBacklog(Level.WARN);
+                return;
+            }
+
+            // Remove bean if it's already registered
+            if (l.getBeanBySystemName(e1.getSystemName()) != null) {
+                l.deregister(e1);
+            }
+            // Remove bean if it's already registered
+            if (l.getBeanBySystemName(e2.getSystemName()) != null) {
+                l.deregister(e2);
+            }
+
+            // Register the bean once. This should be OK.
+            l.register(e1);
+
+            String expectedMessage = "the named bean is registered twice: " + e1.getSystemName();
+            // Register bean twice. This should fail with an IllegalArgumentException.
+            l.register(e1);
+            JUnitAppender.assertWarnMessage(expectedMessage);
+
+            // Use reflection to change the systemName of e2
+            // Try to find the field
+            Field f1 = getField(e2.getClass(), "mSystemName");
```

Comment:
Instead of insisting that NamedBeans have specific non-public implementation details, create a new NamedBean with the existing SystemName.

Reply:
How do I do that? There is no method to create a new bean of the correct type, which is different from the existing registered bean. And I cannot create some other bean, for example a turnout, since the method register() requires that the bean is of the correct type, or I get a compile error.

Reply:
I would think that `E e3 = new E(s1, "e3");` should work to create an E that can be registered (or fail registration because it violates a constraint). I know that technique is used elsewhere where `provide()` is not available from the `Manager<E>`, so I don’t see why it wouldn’t work.

Reply:
I get the compile error:
```
error: unexpected type
            E e2 = new E(s2, "s2");
  required: class
  found:    type parameter E
  where E is a type-variable:
    E extends NamedBean declared in class AbstractManagerTestBase
```